resolve #1952 custom column count

leaumar opened the pull request:
Implemented custom column count

Possible UX shortcomings (marked with TODOs):

- does not apply without a restart
- cannot be set separately for portrait and landscape
- does not apply to 2 other activities that deal with columns that I'm not familiar with

But otherwise it just works and I think the code's clean enough :) I put even the number values in locale strings because langs like chinese and japanese have different number symbols...

![image](https://github.com/user-attachments/assets/db209cd6-0a45-4e0f-8d09-1a18b37cc6a8)
![image](https://github.com/user-attachments/assets/bdadc882-48e6-4fed-93d9-0f08047b7f68)
![image](https://github.com/user-attachments/assets/25926997-832a-45ac-afb5-7eb2ec8d3308)

resolve #1952

TheLastProject commented on app/src/main/res/values/settings.xml:
```diff
@@ -36,6 +36,24 @@
         <item>@string/settings_brown_theme</item>
     </string-array>
 
+    <string-array name="column_count_labels">
+        <item>@string/settings_label_default_column_count</item>
+        <item>@string/settings_label_column_count_1</item>
+        <item>@string/settings_label_column_count_2</item>
+        <item>@string/settings_label_column_count_3</item>
+        <item>@string/settings_label_column_count_4</item>
+        <item>@string/settings_label_column_count_5</item>
+    </string-array>
+
+    <string-array name="column_count_values">
+        <item>@string/settings_symbol_default_column_count</item>
+        <item>1</item>
+        <item>2</item>
+        <item>3</item>
+        <item>4</item>
+        <item>5</item>
+    </string-array>
```
Please be consistent in naming with the other string arrays (they use `x_values` and `x_value_strings` and `@string/_ values` for `strings.xml` (they use variable names like `@string/settings_key_x` and `@string/settings_x`)

TheLastProject commented on app/src/main/res/values/strings.xml:
```diff
@@ -318,6 +318,16 @@
+    <string name="settings_label_column_count_4">4</string>
+    <string name="settings_label_column_count_5">5</string>
+    <string name="setting_key_column_count" translatable="false">pref_column_count</string>
+    <string name="settings_title_column_count">Columns (needs restart)</string>
```
Also here, I'm pretty sure we won't need restarts if we code it right:

```suggestion
    <string name="settings_title_column_count">Columns</string>
```

TheLastProject commented on app/src/main/res/values-nl/strings.xml:
```diff
@@ -260,7 +260,10 @@
+    <string name="settings_category_title_cards">Kaart</string>
+    <string name="settings_category_title_cards_overview">Kaartenoverzicht</string>
+    <string name="settings_label_default_column_count">Automatisch</string>
+    <string name="settings_title_column_count">Kolommen (herstart nodig)</string>
```
I also strongly recommend not translating in the MR yet, it is a lot more back-and-forth and doing stuff twice. Much easier to decide on the strings and then translate it in Weblate later :)

leaumar commented:
* refactored, moved, renamed, added, removed
* applies without restart ✔️ 
* text scaling is off at 4, broken at 5 ❌ 

I'll see if I can spot an obvious problem with the scaling logic

![image](https://github.com/user-attachments/assets/96d02777-949a-431a-8cc2-346fa8ee2e9b)
![image](https://github.com/user-attachments/assets/ec579172-930d-4cf8-96c1-1439b3d006a4)
![image](https://github.com/user-attachments/assets/499f44d5-2a27-4bf6-85f9-d434116d03c4)
![image](https://github.com/user-attachments/assets/3ebfa201-e899-42e7-a8d6-bde09a3090c0)
![image](https://github.com/user-attachments/assets/c611d4a4-8667-46b9-a1a3-3ceb03804402)

leaumar commented:
* dropped the column counts where textview scaling breaks, the textlines logic looks ok to me
* what do you think of these more visual options?

![image](https://github.com/user-attachments/assets/8532c2c3-a7f3-4474-99eb-059861d57feb)

TheLastProject commented on app/src/main/java/protect/card_locker/preferences/Settings.java:
```diff
@@ -92,6 +92,14 @@ public String getColor() {
+        var defaultColumnCount = mContext.getResources().getInteger(R.integer.main_view_card_columns);
+        var columnCountSetting = getString(R.string.setting_key_column_count, defaultSymbol);
+        // the pref may be unset or explicitly set to default
+        return columnCountSetting.equals(defaultSymbol) ? defaultColumnCount : Integer.parseInt(columnCountSetting);
```
I think we should probably, if we can't parse the integer, fall back to the default. It *should* never happen, but Catima crashing at launch with no way to change the setting back, requiring a full data wipe would be horrible.

leaumar replied:
agreed

TheLastProject commented on app/src/main/res/values/strings.xml:
```diff
@@ -318,6 +318,14 @@
+    <string name="settings_column_count">Columns</string>
+    <string name="settings_automatic_column_count">Automatic</string>
+    <string name="settings_key_automatic_column_count" translatable="false">default</string>
+    <string name="settings_column_count_1">1: 💳</string>
+    <string name="settings_column_count_2">2: 💳💳</string>
+    <string name="settings_column_count_3">3: 💳💳💳</string>
```
I understand what you were trying to do, but this may hurt screenreader users and I'm pretty sure some translator will break the translation for this. So let's keep it simple :)

```suggestion
    <string name="settings_column_count_1">1</string>
    <string name="settings_column_count_2">2</string>
    <string name="settings_column_count_3">3</string>
```

leaumar replied:
undone

TheLastProject reviewed (commented):
I mostly just have nitpicks at this point, though I am just looking at the code, I haven't ran it yet.

If you want I can also deal with these last nitpicks myself, whatever has your preference :)

To answer your comments:

> text scaling is off at 4, broken at 5, in portrait mode (all fine in landscape) ❌
   > maybe we just drop 4 and 5? I doubt many people have like 40 cards

I think launching with up to 3 columns as option is a good start, we can extend it later :)

> I'll see if I can spot an obvious problem with the scaling logic

My quick guess would be that it might not be the scaling logic but the padding. But I didn't look into it more yet: https://github.com/CatimaLoyalty/Android/blob/45d2b6206ecfecfc244afa6ee56a2e53f0b04d8c/app/src/main/res/layout/loyalty_card_layout.xml#L46

> what do you think of these more visual options?

I think it's much better to just use the numbers. Visual things like this may cause difficulty for users with screen readers and translators may mistranslate it, copy-pasting it in incorrect counts or with broken emojis.

leaumar commented on app/src/main/java/protect/card_locker/MainActivity.java:
```diff
@@ -357,6 +359,12 @@ protected void onResume() {
             mBarcodeScannerLauncher.launch(intent);
         });
         addButton.bringToFront();
+
+        GridLayoutManager layoutManager = (GridLayoutManager) mCardList.getLayoutManager();
+        if (layoutManager != null) {
+            var settings = new Settings(this);
+            layoutManager.setSpanCount(settings.getPreferredColumnCount());
+        }
```
I'll leave this to you since it's out of my scope

leaumar commented on app/src/main/java/protect/card_locker/CardShortcutConfigure.java:
```diff
@@ -48,16 +49,24 @@ public void onCreate(Bundle bundle) {
-        GridLayoutManager layoutManager = (GridLayoutManager) cardList.getLayoutManager();
-        if (layoutManager != null) {
-            layoutManager.setSpanCount(getResources().getInteger(R.integer.main_view_card_columns));
-        }
 
         Cursor cardCursor = DBHelper.getLoyaltyCardCursor(mDatabase, DBHelper.LoyaltyCardArchiveFilter.All);
         mAdapter = new LoyaltyCardCursorAdapter(this, cardCursor, this, null);
         cardList.setAdapter(mAdapter);
     }
 
+    @Override
+    protected void onResume() {
+        super.onResume();
+
+        RecyclerView cardList = binding.list;
+        GridLayoutManager layoutManager = (GridLayoutManager) cardList.getLayoutManager();
+        if (layoutManager != null) {
+            var settings = new Settings(this);
+            layoutManager.setSpanCount(settings.getPreferredColumnCount());
+        }
+    }
+
```
inlined

leaumar commented on app/src/main/java/protect/card_locker/LoyaltyCardCursorAdapter.java:
```diff
@@ -123,6 +123,7 @@ public void onBindViewHolder(LoyaltyCardListItemViewHolder inputHolder, Cursor i
         }
 
         inputHolder.mCardIcon.setContentDescription(loyaltyCard.store);
+        // having more than 3 columns in portrait causes the cardText to shrink excessively relative to the available space
```
fixed with dynamic padding

leaumar commented on app/src/main/java/protect/card_locker/preferences/Settings.java:
```diff
@@ -92,6 +100,22 @@ public String getColor() {
 
+    public int getPreferredColumnCount() {
+        var orientation = mContext.getResources().getConfiguration().orientation;
+        var defaultColumnCount = orientation == ORIENTATION_PORTRAIT ? 2 : 3;
```
I thought about the duplication of these values between Settings and the pref `defaultValue` attrs, but if you make them xml values you have to duplicate them between `column_count_values` and `@string/setting_key_default_column_count_$orient` instead, so that hardly seems better

TheLastProject replied:
I disagree about this change. Because you're changing the logic.

The old logic was as follows:
By default, use 2 columns (https://github.com/CatimaLoyalty/Android/blob/main/app/src/main/res/values/integers.xml)
If the screen width is at least 600dp, use 3 colums (https://github.com/CatimaLoyalty/Android/blob/main/app/src/main/res/values-w600dp/integers.xml)

For most devices, this tends to roughly match portrait vs. landscape. However, it's not consistently the same. For example, [square phones exist](https://github.com/CatimaLoyalty/Android/issues/1339) and Android flipphones with a small secondary screen (which would probably run Catima very well!) are existing too now. Certain tablets and PC devices may, even in landscape, still be at least 600dp.

Can you move it out of the values file to code? Perhaps, although I don't see a benefit from it, the XML is a good location for it. Can you remove the "Automatic" setting and just hardcode it on 2 or 3? I don't think so, not without breaking some use cases.

TheLastProject reviewed (approved):
I'll do a bit more messing around to see if I can break it but otherwise I'll merge this soon :)
